### PR TITLE
[DomCrawler] add a value() method, normalize whitespaces

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21302, #23626
| License       | MIT
| Doc PR        | TODO

### CODE REVIEW

File: src/Symfony/Component/DomCrawler/Tests/CrawlerTest.php [hunk 1]
```diff
@@ -382,16 +382,35 @@ public function testNodeName()
...
+    }
+
+    /**
+     * @expectedException \InvalidArgumentException
+     */
+    public function testValue()
+    {
+        $this->assertEquals("one   two\nthree   ", $this->createTestCrawler()->filterXPath('//div[@id="text-whitespaces"]')->value(), '->value() returns the node value of the first element of the node list');
+
+        $this->createTestCrawler()->filterXPath('//ol')->value();
```
Comment (Tobion) [inline src/Symfony/Component/DomCrawler/Tests/CrawlerTest.php]: this should be a separate test like `testValueThrowsExceptionWhenNodeListEmpty`. if the above call aleady fails with InvalidArgumentException, the test would still pass.

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 2]
```diff
@@ -556,13 +556,37 @@ public function nodeName()
...
+    public function text($normalizeWhitespaces = false)
+    {
+        if (!$normalizeWhitespaces) {
+            @trigger_error(sprintf('Not normalizing whitespace characters when using the %s() method is deprecated since version 3.4. The %s() method will return the first node\'s value with whitespaces being normalized in Symfony 4.0. Use the value() method instead if you do not want to have whitespaces normalized.', __METHOD__, __METHOD__), E_USER_DEPRECATED);
+        }
+
+        $value = $this->value();
+
+        if ($normalizeWhitespaces) {
+            $value = trim(preg_replace('/\s++/', ' ', $value));
```
Comment (vlakoff) [inline src/Symfony/Component/DomCrawler/Crawler.php]: Isn't there an extraneous `+` ?

**Comment by vlakoff** [comment]
So, on 3.4 people will have to update their codebase to `text(true)`, then in 4.0 they will have to update it again... We're not very kind to them.

Couldn't we drop the 3.4 step and just make the 4.0 change? I don't really see the benefits of the 3.4 step, apart from adding work.

**Comment by xabbuh** [comment]
There is no need to update the code in 4.0 again. You *could* omit the argument then, but it's not necessary.

File: UPGRADE-3.4.md [hunk 3]
```diff
@@ -63,6 +63,12 @@ Debug
 
  * Support for stacked errors in the `ErrorHandler` is deprecated and will be removed in Symfony 4.0.
 
+DomCrawler
+----------
+
+ * added an argument to the `text()` method to opt-in normalizing whitespaces, not passing `true` is deprecated, for the
+   previous behavior of the `text()` method, use the new `value()` method
```
Comment (ro0NL) [inline UPGRADE-3.4.md]:
```
 * Added an argument to the `text()` method to opt-in normalizing whitespaces, not passing `true` is deprecated. For the
   previous behavior of the `text()` method, use the new `value()` method.
```
? (see dots)

File: UPGRADE-4.0.md [hunk 4]
```diff
@@ -186,6 +186,11 @@ DependencyInjection
 
  * The `ExtensionCompilerPass` has been moved to before-optimization passes with priority -1000.
 
+DomCrawler
+----------
+
+ * the `text()` method now normalizes whitespaces of the node text value
```
Comment (ro0NL) [inline UPGRADE-4.0.md]: same. Compared to above/below blocks this should be sentences no?

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 2 above)
Comment (ro0NL) [reply src/Symfony/Component/DomCrawler/Crawler.php]: see http://www.regular-expressions.info/possessive.html Basically _repeats the token as many times as possible_, thus optimized by default.

Comment (vlakoff) [reply src/Symfony/Component/DomCrawler/Crawler.php]: TIL, thanks :)
Though I'm not sure it makes a difference here, as there is nothing after, thus no backtracking.

**Review by ro0NL — APPROVED** [review]

**Review by javiereguiluz — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
I'm sorry I don't get the migration path at all.
I understand we're fighting adding a flag... By adding a flag...
As far as the migration path is concerned, this will require ppl to change their code twice: once when moving to 4.0, by adding a flag (failed goal already)
And twice in 4.1 to remove the deprecation...

If we must add a flag because there is no other migration path, let's settle on having a flag and keep it. This would means zero deprecation (but for ppl extending the class, but that's edgy.)

**Comment by nicolas-grekas** [comment]
Oh, and we could then, in 4.1, deprecate the flag and replace it by this value() method.
That'd mean *one* deprecation, thus one migration.
Better, isn't it?

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 5]
```diff
@@ -556,13 +556,37 @@ public function nodeName()
...
+     * @param bool $normalizeWhitespaces Pass true to normalize whitespaces, not passing true is deprecated since version 3.4.
+     *
+     * @return string The node value
+     *
+     * @throws \InvalidArgumentException When current node is empty
+     */
+    public function text($normalizeWhitespaces = false)
+    {
+        if (!$normalizeWhitespaces) {
+            @trigger_error(sprintf('Not normalizing whitespace characters when using the %s() method is deprecated since version 3.4. The %s() method will return the first node\'s value with whitespaces being normalized in Symfony 4.0. Use the value() method instead if you do not want to have whitespaces normalized.', __METHOD__, __METHOD__), E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DomCrawler/Crawler.php]: This deprecation is not required and should be removed.
Also, the added arg is a bc break right now, should be done using func_get_arg()

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]

**Comment by xabbuh** [comment]
Fair enough, I reverted the deprecation.

Status: Needs Review

**Review by nicolas-grekas — COMMENTED** [review]
That's still strange, sorry :)
What about deprecating `test()`, and add a `$normalizeWhitespace = true` flag to `value()`
The flag is a requirement to me, I can't find any better path that would be migration-path-friendly.
But by defaulting to "true", value could just be good by default, WDYT?

**Comment by xabbuh** [comment]
I understand what you mean, but if we did that, `value()` would have to get passed `false` to behave as I would have expected. I mean the name indicates returning the raw value, doesn't it? So if we want to go the way with the flag, I suggest to just add it to `text()` and then decide somewhere along the 4.x if we deprecate the flag and introduce something else or just live with the flag.

**Comment by stof** [comment]
I the context of an HTML crawler, I think that `value()` would be confusing it. If I have a crawler holding an input element or a select option, I would understand `->value()` as returning the value of the input (which is in an attribute), not the inner text (which is what `->nodeValue` does).
FYI, the JS property corresponding to this property is `textContent`.

The normalization you are doing here for `text` corresponds to *part* of the behavior non-standard `innerText` property of JS DOM APIs (we cannot reproduce the full behavior, as this involves removing text from non-rendered children and we are not doing rendering, and reproducing other parts cannot be done with a simple regex: https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute)

**Comment by xabbuh** [comment]
I suggest to postpone this PR for Symfony 4.1 to not rush anything into core with which we may not be completely satisfied.

**Comment by fabpot** [comment]
@xabbuh Time to resurrect this PR?

**Comment by vlakoff** [comment]
To avoid all that upgrade path mess with the method parameters, I'd suggest just adding a new method as I proposed initially.

**Comment by Simperfit** [comment]
@xabbuh do you want me to finish this one ?

**Comment by xabbuh** [comment]
@Simperfit If you have some time, please feel free to take over. 👍

**Comment by fabpot** [comment]
@Simperfit Is this PR still on your radar?

**Comment by Simperfit** [comment]
@fabpot doing it today

Le lun. 8 juil. 2019 à 12:07, Fabien Potencier <notifications@github.com> a
écrit :

**Comment by xabbuh** [comment]
closing in favour of #32440

**Comment by xabbuh** [comment]
Thank you for taking over @Simperfit!

**Pull request closed without merging**